module/mempool/herocache/backdata/heropool/pool_test.go

```diff
@@ -215,6 +217,111 @@ func TestInvalidateEntity(t *testing.T) {
+func TestAddAndRemoveEntities(t *testing.T) {
+	for _, tc := range []struct {
+		limit               uint32       // capacity of the pool
+		entityCount         uint32       // total entities to be stored
+		ejectionMode        EjectionMode // ejection mode
+		numberOfOperations  int
+		probabilityOfAdding float32
+	}{
+		{
+			limit:               500,
+			entityCount:         1000,
+			ejectionMode:        LRUEjection,
+			numberOfOperations:  1000,
+			probabilityOfAdding: 0.8,
+		},
+		{
+			limit:               500,
+			entityCount:         1000,
+			ejectionMode:        NoEjection,
+			numberOfOperations:  1000,
+			probabilityOfAdding: 0.8,
+		},
+		{
+			limit:               500,
+			entityCount:         1000,
+			ejectionMode:        RandomEjection,
+			numberOfOperations:  1000,
+			probabilityOfAdding: 0.8,
+		},
+	} {
+		t.Run(fmt.Sprintf("%d-limit-%d-entities", tc.limit, tc.entityCount), func(t *testing.T) {
+			testAddRemoveEntities(t, tc.limit, tc.entityCount, tc.ejectionMode, tc.numberOfOperations, tc.probabilityOfAdding)
+		})
+	}
+}
+
+// Indexes contains an index in the pool of an inserted entity and an index of this entity
+// in the array of entities created by the EntityListFixture.
+type Indexes struct {
+	indexInPool        EIndex
+	indexInEntitiesArr int
+}
+
+func testAddRemoveEntities(t *testing.T, limit uint32, entityCount uint32, ejectionMode EjectionMode, numberOfOperations int, probabilityOfAdding float32) {
+	// create and log the seed.
+	testSeed := time.Now().UnixNano()
+	rand.Seed(testSeed)
+	fmt.Println("Seed used for this testAddRemoveEntities execution", testSeed)
+
+	pool := NewHeroPool(limit, ejectionMode)
+
+	entities := unittest.EntityListFixture(uint(entityCount))
+
+	// an array of random owner Ids.
+	ownerIds := make([]uint64, entityCount, entityCount)
+	// generate ownerId to index in the entities arrray.
+	for i := 0; i < int(entityCount); i++ {
+		ownerIds[i] = rand.Uint64()
+	}
+	// this map maintains entities currently stored in the pool.
+	var addedEntities map[flow.Identifier]Indexes
+	addedEntities = make(map[flow.Identifier]Indexes)
+
+	for i := 0; i < numberOfOperations; i++ {
+		// choose between Add and Remove with a probability of probabilityOfAdding and 1-probabilityOfAdding respectively.
+		if rand.Float32() < probabilityOfAdding || len(addedEntities) == 0 {
+			// adding an entity
+			entityToAdd := rand.Intn(int(entityCount))
+			// check that entity is not already inserted.
+			_, found := addedEntities[entities[entityToAdd].ID()]
+			if !found {
+				indexInThePool, _, ejectedEntity := pool.Add(entities[entityToAdd].ID(), entities[entityToAdd], uint64(ownerIds[entityToAdd]))
+				if indexInThePool != InvalidIndex {
+					addedEntities[entities[entityToAdd].ID()] = Indexes{indexInThePool, entityToAdd}
+				}
+				if ejectedEntity != nil {
+					delete(addedEntities, ejectedEntity.ID())
+				}
+			}
```

Comment:
When the test adds an entity, it must perform a sanity check that the same entity is retrievable using `pool.Get` method. See an example [here](https://github.com/onflow/flow-go/blob/master/module/mempool/herocache/backdata/heropool/pool_test.go#L618-L622).

Reply:
I, actually, had a check at the end of the test which would use ```addedEntities``` and Get of the pool.But now indeed I no longer see it. Probably, I accidentally removed it while implementing this version. But as I understand you now you suggest to perform a check after each operation. Is that correct ?